guides/building-a-middleware-from-scratch.md

```diff
@@ -0,0 +1,693 @@
+    inner: T,
+    timeout: Duration,
+}
+```
+
+As we learned in ["Inventing the `Service` trait"][invent] its important for
+services to implement `Clone` such that you can convert the `&mut self` given to
+`Service::call` into an owned `self` that can be moved into the response future,
+if necessary. We should therefore add `#[derive(Clone)]` to our struct. We
+should also drive `Debug` while we're at it:
+
+```rust
+#[derive(Debug, Clone)]
+struct Timeout<S> {
+    inner: S,
+    timeout: Duration,
+}
+```
+
+Next we write a constructor:
+
+```rust
+impl<S> Timeout<S> {
+    pub fn new(inner: S, timeout: Duration) -> Self {
+        Timeout { inner, timeout }
+    }
+}
+```
+
+Note that we don't have to add any trait bounds on `S` even though we expect `S`
+to implement `Service`. In Rust is common practice to only add trait bounds
+where they're required and since `new` doesn't require anything from `S` we
+don't add any trait bounds.
+
+Now the interesting bit. How to implement `Service` for `Timeout<S>`? Lets start
+with an implementation that just forwards everything to the inner service:
+
+```rust
+use tower::Service;
+use std::task::{Context, Poll};
+
+impl<S, Request> Service<Request> for Timeout<S>
+where
+    S: Service<Request>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // Our middleware doesn't care about backpressure so its ready as long
+        // as the inner service is ready.
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: Request) -> Self::Future {
+        self.inner.call(request)
+    }
+}
+```
+
+Until you've written lots of middleware writing out a skeleton like this makes
+the process a bit easier.
+
+To actually add a timeout to the inner service what we essentially have to do is
+detect when the future returned by `self.inner.call(request)` has been running
+longer than `self.duration` and abort with an error.
+
+The approach we're going to take is to call [`tokio::time::sleep`] to get a
+future that completes when we're out of time and then select the value from
+whichever of the two futures is the first to complete. We could also use
+`tokio::time::timeout` but `sleep` works just as well.
+
+Creating both futures is done like this:
+
+```rust
+use tokio::time::sleep;
+
+fn call(&mut self, request: Request) -> Self::Future {
+    let response_future = self.inner.call(request);
+
+    // This variable has type `tokio::time::Sleep`.
+    //
+    // We don't have to clone `self.duration` as it implements the `Copy` trait.
+    let sleep = tokio::time::sleep(self.timeout);
+
+    // what to write here?
+}
+```
+
+One possible return type is `Pin<Box<dyn Future<...>>>`. However we want our
+`Timeout` to add as little overhead as possible, so we would like to find a way
+to avoid allocating a `Box`. Imagine we have a large stack, with dozens of
```

Comment:
Would be good to call out that `type_alias_impl_trait` fixes this issue since we can return an unnamed type.

Reply:
I've added a footnote about it. Generally not a fan of mentioning features that have been "coming soon" for years 😜